packages/excalidraw/fractionalIndex.ts

```diff
@@ -0,0 +1,266 @@
+import { mutateElement } from "./element/mutateElement";
+import { ExcalidrawElement } from "./element/types";
+import {
+  generateKeyBetween as _generateKeyBetween,
+  generateJitteredKeyBetween as _generateJitteredKeyBetween,
+  generateNKeysBetween as _generateNKeysBetween,
+  generateNJitteredKeysBetween as _generateNJitteredKeysBetween,
+  base62CharSet as _base62CharSet,
+  indexCharacterSet,
+} from "fractional-indexing-jittered";
+import { ENV } from "./constants";
+import { Mutable } from "./utility-types";
+
+type FractionalIndex = ExcalidrawElement["fractionalIndex"];
+
+const base36CharSet = indexCharacterSet({
+  chars: "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ",
+  firstPositive: "A",
+  mostPositive: "Z",
+});
+
+const { chars, firstPositive, mostPositive, mostNegative } = _base62CharSet();
+
+const base62CharSet = indexCharacterSet({
+  chars,
+  firstPositive,
+  mostPositive,
+  mostNegative,
+  jitterRange: 6,
+});
+
+const charSet =
+  import.meta.env.DEV || import.meta.env.MODE === ENV.TEST
+    ? base36CharSet
+    : base62CharSet;
+
+export const generateKeyBetween = (
+  lower: string | null,
+  upper: string | null,
+) => _generateKeyBetween(lower, upper, charSet);
+
+export const generateJitteredKeyBetween = (
+  lower: string | null,
+  upper: string | null,
+) => _generateJitteredKeyBetween(lower, upper, charSet);
+
+export const generateNKeysBetween = (
+  lower: string | null,
+  upper: string | null,
+  n: number,
+) => _generateNKeysBetween(lower, upper, n, charSet);
+
+export const generateNJitteredKeysBetween = (
+  lower: string | null,
+  upper: string | null,
+  n: number,
+) => _generateNJitteredKeysBetween(lower, upper, n, charSet);
+
+export const orderByFractionalIndex = (allElements: ExcalidrawElement[]) => {
+  return allElements.sort((a, b) => {
+    if (a.fractionalIndex && b.fractionalIndex) {
+      if (a.fractionalIndex < b.fractionalIndex) {
+        return -1;
+      } else if (a.fractionalIndex > b.fractionalIndex) {
+        return 1;
+      }
+      return compareStrings(a.id, b.id);
+    }
+
+    // respect the order of the array
+    return 1;
+  });
+};
+
+export const validateFractionalIndices = (
+  elements: readonly ExcalidrawElement[],
+) => {
+  for (let i = 0; i < elements.length; i++) {
+    const element = elements[i];
+
+    if (!element.fractionalIndex) {
+      return false;
+    }
+
+    const successor: ExcalidrawElement | undefined = elements[i + 1];
+
+    if (
+      successor?.fractionalIndex &&
+      element.fractionalIndex > successor.fractionalIndex
```

Comment:
@Mrazator here should be `>=` here instead, as we would like to catch also equalities in the future. Potentially could be also unified with `isValidFractionalIndex` below.

Reply:
Solved differently as part of #7359, we make sure indices are always deduplicated.